Sources/MongoSwift/ChangeStream.swift

```diff
@@ -0,0 +1,165 @@
+    let cursor = self.changeStream
+    // Allocate space for a reference to a BSON pointer.
+    let out = UnsafeMutablePointer<BSONPointer?>.allocate(capacity: 1)
+    defer {
+        out.deinitialize(count: 1)
+        out.deallocate()
+    }
+
+    // Check that there’s another document in the stream.
+    guard mongoc_change_stream_next(cursor, out) else {
+      return nil
+    }
+
+    guard let pointee = out.pointee else {
+        fatalError("Could not advance the change stream.")
+    }
+
+    // we have to copy because libmongoc owns the pointer.
+    let doc = Document(copying: pointee)
+
+    // Update the resumeToken with the `_id` field from the document.
+    if let resumeToken = doc["_id"] as? Document {
+        self.resumeToken = ChangeStreamToken(resumeToken: resumeToken)
+    } else {
+        self.swiftError = RuntimeError.internalError(message: "_id field is missing from the change stream document.")
+        return nil
+    }
+
+    do {
+        return try decoder.decode(T.self, from: doc)
+    } catch {
+        self.swiftError = error
+    }
+
+    return nil
+  }
+
+  /**
+   * Returns the next `T` in this change stream or `nil`, or throws an error if one occurs -- compared to `next()`,
+   * whichreturns `nil` and requires manually checking for an error afterward. Will block for a maximum of
+   * `maxAwaitTimeMS` milliseconds as specified in the `ChangeStreamOptions`, or for the server default timeout if
+   * omitted.
+   * - Returns: the next `T` in this change stream, or `nil` if at the end of the change stream cursor.
+   * - Throws:
+   *   - `ServerError.commandError` if an error occurs on the server while iterating the change stream cursor.
+   *   - `UserError.logicError` if this function is called and the session associated with this change stream is
+   *     inactive.
+   *   - `DecodingError` if an error occurs decoding the server's response.
+   */
+  public func nextOrError() throws -> T? {
+    if let next = self.next() {
+      return next
+    }
+
+    if let error = self.error {
+      throw error
+    }
+    return nil
+  }
+
+  /**
+   * Initializes a `ChangeStream`.
+   * - Throws:
+   *   - `ServerError.commandError` if an error occurred on the server when creating the `mongoc_change_stream_t`.
+   *   - `UserError.invalidArgumentError` if the `mongoc_change_stream_t` was created with invalid options.
+   */
+  internal init(stealing changeStream: OpaquePointer,
+                client: MongoClient,
+                session: ClientSession? = nil,
+                decoder: BSONDecoder) throws {
+    self.changeStream = changeStream
+    self.session = session
+    self.client = client
+    self.decoder = decoder
+    self.resumeToken = ChangeStreamToken(resumeToken: [])
+
+    if let err = self.error {
+      throw err
+    }
+  }
+
+  /// Cleans up internal state.
+  deinit {
+    guard let cursor = self.changeStream else {
+        return
+    }
+    mongoc_change_stream_destroy(changeStream)
```

Comment:
I think we can just directly pass in `self.changeStream` here and we don't need the guard above.

Reply:
nit: use `self.changeStream`. swift doesn't require explicit `self` but we try to always use it. (there's a swiftlint rule for this but it misses most violations 🤦‍♀️